Fix - Trend charts look very broken with new dashboard grid changes

kamilmielnik opened the pull request:
Closes #31628

### Description

This fixes overflow in:
- simple scalar charts (screenshots: [before](https://github.com/metabase/metabase/assets/6830683/d5d121fb-8f0d-4bce-a4e7-955ae6545256) / [after](https://github.com/metabase/metabase/assets/6830683/1d46ff18-76d7-4666-8f41-b6cac3eb68ab))
- trend charts (screenshots: [before](https://github.com/metabase/metabase/assets/6830683/fba4ee07-a93d-4aaf-bb84-f37fe6bcfcc0) / [after](https://github.com/metabase/metabase/assets/6830683/3125ac69-78a6-4901-8014-15ca345552b2))

It works down to 2x2 size in `smartscalar`, so this PR also updates the minimum allowed size of `smartscalar`.

To see how the regression looked like please [watch these recordings](https://github.com/metabase/metabase/issues/31628#issuecomment-1597120018).

Related Slack threads:
- https://metaboat.slack.com/archives/C057T1QTB3L/p1687167736234729
- https://metaboat.slack.com/archives/C505ZNNH4/p1687463683427169

### Tests

I went with e2e tests because in unit tests elements don't have dimensions and mocking the visual state that we expect would defy the purpose.

### How to verify

Create a dashboard and add a bunch of questions visualized as: "number" and "trend" (here's the question for the trend chart I'm using in my recordings: [screenshot](https://github.com/metabase/metabase/assets/6830683/d14143a9-61c2-4b6f-8bfe-31001484f523)).

The following things affect how the cards are displayed, please try various combinations:
- card size
- viewport size (browser window size)
- whether left-hand side navigation sidebar is open or not
- whether the card has a description or not (there will be an "i" icon displayed)
- size of the text values displayed in the card (card value, card name, trend difference, etc.)

Less important but also affecting:
- font (user setting)
- language (user setting)

How it's meant to work:
- card title will be displayed in 2 lines if possible, in 1 line otherwise
- card value & title will be truncated with ellipsis if there's not enough horizontal space
- card value font size is automatically computed based on the available horizontal & vertical space (there is a min & max to this)
- if something is truncated with ellipsis, a tooltip with a complete value should be shown on hover
- if something is not truncated with ellipsis, there should be no tooltip
- description icon is shown only when description is present and description is shown in tooltip on hover
- in trend cards:
  - the `"was XYZ last month"` label will be displayed only if it fits horizontally - otherwise it will be shown in tooltip on hover
    - except on mobile resolutions, where it will be displayed in a separate row
  - the % difference will be always shown
    - it will try to use up to 2 decimal points precision if possible - otherwise it will try to use less precision
    - if there's not enough room to show it anyway - it will be truncated
    - icon is always shown

### Demo


#### Trend - Before

https://github.com/metabase/metabase/assets/6830683/5ffbd7a2-f95a-4e40-90b9-79656751330b

#### Trend - After

https://github.com/metabase/metabase/assets/6830683/8ed64ce7-baaf-48a6-8274-b81114f3b41c

#### Number - Before

https://github.com/metabase/metabase/assets/6830683/e06da892-6e07-4bd4-8cb1-51d699fe43c3




#### Number - After

https://github.com/metabase/metabase/assets/6830683/0c41745f-fb86-4fb7-b193-bcd642d678ec

kamilmielnik committed:
Fix component name

kamilmielnik committed:
Show only 1 title line on small SmartScalars

kamilmielnik committed:
Style PreviousValueVariation and the separator
- Get rid of flexbox so that the default browser text-wrapping algorithm is used
- Be pixel-perfect about margin-left

kamilmielnik committed:
Bring back height calculation based on number of lines

kamilmielnik committed:
Format change

kamilmielnik committed:
Fix ellipsis in ScalarValue

kamilmielnik committed:
Make spacing consistent

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

kamilmielnik committed:
Style VariationIcon and VariationValue
- get rid of a utility class and inline CSS

kamilmielnik committed:
Add ellipsis to previous value, style tooltip

kamilmielnik committed:
Ensure consistent width of SmartScalar children

kamilmielnik committed:
Update JS to match CSS

kamilmielnik committed:
Refactor measureText to return the whole TextMetrics object

kamilmielnik committed:
Extract hardcoded padding out of ScalarValue

kamilmielnik committed:
Update measure text to return height explicitly

kamilmielnik committed:
Account for height in the findSize function

kamilmielnik committed:
Improve spacing

kamilmielnik committed:
Improve styling

kamilmielnik committed:
Fix build, bring back removed styling

kamilmielnik committed:
Make previous value responsive

kamilmielnik committed:
Revert #31724

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

kamilmielnik committed:
Fix typing
- Introduce TextWidthMeasurer type

kamilmielnik committed:
Remove redundant CSS and elements

kamilmielnik committed:
Rename formatChangeSmart to formatChangeAutoPrecision

kamilmielnik committed:
Fix ellipsis on previous value

kamilmielnik committed:
Improve 2 lines threshold value and introduce explanatory constant

kamilmielnik committed:
Format code

kamilmielnik committed:
Remove redundant code

kamilmielnik committed:
Update findSize tests

kamilmielnik commented on frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx:
```diff
@@ -145,17 +195,51 @@ export default class Smart extends Component {
+      previousValue,
+      settings.column(column),
     );
+    const disabledSeparator = ""; // avoid creating new translation key
+    const previousValueContent = jt`${disabledSeparator} was ${previousValueDisplay} ${granularityDisplay}`;
```
I'm not sure why separator was originally included in the translation.

Please let me know if I should introduce a new translation key.

kamilmielnik committed:
Fix overflow in Scalars

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

kamilmielnik committed:
Add overflow test boilerplate

kamilmielnik committed:
Include more sizes in the test

kamilmielnik committed:
Format code

kamilmielnik committed:
Add tests for scalar
- add scalar-description to assertions

kamilmielnik committed:
Extract assertDescendantNotOverflowsContainer

kamilmielnik committed:
Remove redundant check

kamilmielnik committed:
Extract setupDashboardWithQuestionInCards

kamilmielnik committed:
Reorder cards from biggest to lowest horizontally

kamilmielnik committed:
Extract assertDescendantsNotOverflowDashcards

kamilmielnik committed:
Add test case for when sidebar is open

kamilmielnik committed:
Format code

kamilmielnik committed:
Ignore elements without dimensions in assertions

kamilmielnik committed:
Reorder cards from biggest to smallest vertically

kamilmielnik committed:
Add 1x cards in tests for scalars

kamilmielnik committed:
Add tests for various viewports

kamilmielnik committed:
Improve naming

kamilmielnik commented on e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:
```diff
@@ -0,0 +1,189 @@
+  });
+};
+
+const assertDescendantNotOverflowsContainer = (descendant, container) => {
```
This function is generic and could be used in other tests. Do you think I should make it a helper or a command?

kamilmielnik committed:
Format code

kamilmielnik committed:
Fix cursor: pointer being shown on container hover instead of icon hover

kamilmielnik committed:
Improve vertical icon alignment, make ScalarDescriptionPlaceholder padding consistent with ScalarDescriptionContainer

kamilmielnik committed:
Account for title lines count in Scalar

kamilmielnik committed:
Refactor tests to allow better test case descriptons

kamilmielnik committed:
Use useful viewports

kamilmielnik committed:
Find a testable way to fix ellipsis

kamilmielnik committed:
Adjust test data

kamilmielnik committed:
Add assertion messages

kamilmielnik committed:
Fix tooltip not shown on value when ellipsis is applied

kamilmielnik committed:
Add test for tooltips in 1x2 card

kamilmielnik committed:
Add tests for 2x2 & 5x3 cards
- fix CSS issues breaking tests

kamilmielnik committed:
Add tests for 2x2 card

kamilmielnik committed:
Add tests for more card sizes

kamilmielnik committed:
Assert ellipsis is applied

kamilmielnik committed:
Fix assertions

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

kamilmielnik committed:
Update unit tests

WiNloSt commented on frontend/src/metabase/visualizations/components/ScalarValue/utils.unit.spec.ts:
```diff
@@ -44,14 +45,15 @@ describe("findSize", () => {
-      .mockImplementationOnce(createMockMeasureText(110))
-      .mockImplementationOnce(createMockMeasureText(100)) // this is the one we want
-      .mockImplementationOnce(createMockMeasureText(90));
+      .mockImplementationOnce(createMockMeasureText(120, 50))
+      .mockImplementationOnce(createMockMeasureText(110, 50))
+      .mockImplementationOnce(createMockMeasureText(100, 50)) // this is the one we want
+      .mockImplementationOnce(createMockMeasureText(90, 50));
```
nit: I think to somewhat show the actual behavior you should shrink the height proportionally to the width, since `findSize` will use the same text but with reduced `size` each time. So you could get a result that will be narrower and shorter each time.

WiNloSt commented on frontend/src/metabase/visualizations/components/ScalarValue/utils.unit.spec.ts:
```diff
@@ -61,11 +63,61 @@ describe("findSize", () => {
+
+  it("returns the first size with which text height is smaller than the target height", () => {
+    measureTextSpy
+      .mockImplementationOnce(createMockMeasureText(50, 120))
+      .mockImplementationOnce(createMockMeasureText(50, 110))
+      .mockImplementationOnce(createMockMeasureText(50, 100)) // this is the one we want
+      .mockImplementationOnce(createMockMeasureText(0, 90));
```
Similiar to my previous comment, I think we should kind of proportionally, mock the width and height. They don't need to maintain the exact aspect ratio, but at least they need to shrink together.

Also, I'm not sure if the last line that sets the width to `0` is a typo or not.

kamilmielnik replied:
> Also, I'm not sure if the last line that sets the width to 0 is a typo or not.

Yes, good catch :+1: 
f0219dd4d85a2c4e01ac6d71bcf2ed1dd0db2505

kamilmielnik replied:
c88f7c77f9b474a5c28325b3a74d314f56cc6b10

WiNloSt reviewed (commented):
This is more than half of the full review. I'll continue reviewing this afterward.

kamilmielnik commented on frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx:
```diff
@@ -23,21 +23,26 @@ interface ScalarValueWrapperProps {
 }
 
 export const ScalarValueWrapper = styled.h1<ScalarValueWrapperProps>`
+  display: inline;
```
Ellipsis does not work without it

kamilmielnik committed:
Rename TITLE_2_LINES_THRESHOLD to TITLE_2_LINES_HEIGHT_THRESHOLD

kamilmielnik commented on frontend/src/metabase/visualizations/visualizations/Scalar.jsx:
```diff
@@ -29,6 +29,26 @@ function legacyScalarSettingsToFormatOptions(settings) {
+const TITLE_ICON_SIZE = 10;
+const SCALAR_TITLE_LINE_HEIGHT = 23;
+
+const TITLE_2_LINES_THRESHOLD = 120; // determined empirically
```
823ed32d14e7232b4c2378f013f45be2aba7d790

kamilmielnik committed:
Fix typo

kamilmielnik committed:
Make mock data reflect real data a bit more

kamilmielnik committed:
Fix cursor: pointer being applied on the container hover instead of icon

kamilmielnik commented on frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx:
```diff
@@ -54,14 +59,18 @@ export const ScalarTitleContent = styled.h3`
+`;
+
+export const ScalarDescriptionIcon = styled(Icon)`
+  &:hover {
+    color: ${color("brand")};
+  }
```
Actually, I forgot to move `cursor: pointer`. Fixed: 27152d77b98d036fc39e237a694d091446bfb9ba

WiNloSt replied:
That's clear. Thanks.

WiNloSt commented on frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx:
```diff
@@ -54,14 +59,18 @@ export const ScalarTitleContent = styled.h3`
 
 export const ScalarDescriptionPlaceholder = styled.div`
   width: 1.5rem;
+  margin-top: 0.2rem;
+  padding-right: 0.5rem;
```
There isn't an RFC about this yet, but this is a part of introducing Metabase components or what we've been doing in the project UI library. [Here's some of the conversation](https://metaboat.slack.com/archives/C057WD5L0JG/p1686227773463059?thread_ts=1686219268.256349&cid=C057WD5L0JG), and [a PR that starts using `Box`](https://github.com/metabase/metabase/pull/31426) component in our app (the description is outdated though, nothing there is about `createStyles`)

I think the thing we're trying to avoid is to stop positioning elements inside the element itself if that makes sense. Generally, I view a component as a standalone UI without either margin or padding. And the parent components that want to use any specific UI need to know where to position those UI themselves that's why I think leaving spacing e.g. `margin` and `padding` in the parent component would make more sense.

This will enable us to use reusable components with less friction as well, previously we need to create a new `*.styled.*` file if it doesn't exist. Extend the component you want to use with `styled(YourComponent)` then add one or 2 lines for the spacing. That won't be necessary with `Box` and `Flex` and all other layout components Mantine provide anymore.

kamilmielnik replied:
Thank you for a detailed explanation! :100: 

If you don't mind I will skip this until we establish a convention.

WiNloSt replied:
Sounds fine.

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

ranquild commented on frontend/src/metabase/visualizations/visualizations/Scalar.jsx:
```diff
@@ -29,6 +29,27 @@ function legacyScalarSettingsToFormatOptions(settings) {
+const getTitleLinesCount = height =>
+  height > TITLE_2_LINES_HEIGHT_THRESHOLD ? 2 : 1;
+
+const getTitleHeight = ({ isDashboard, showSmallTitle, titleLinesCount }) => {
```
wdyt about moving these utilities to a different (utility) file? I'd also recommend taking a look at how static viz charts (e.g. RowChart) are implemented where there are utility functions that return props for "view" components. Here we could have a function that returns values for props for inner components so there won't be any computation in the component; and we could also test these utility functions

kamilmielnik replied:
With pleasure :rocket: 

07038432ac74e5f7bad50a7e957dbfa66de149b5 + d502456f3743789e72c5179fee5896180902c5bd

kamilmielnik replied:
> and we could also test these utility functions

I didn't add any tests for them. There are 2 types of functions that I managed to extract:
- trivial ones involving arithmetic, which I think don't deserve tests (TS seems enough of a check)
- those using `measureText` - it'd be awkward to unit-test them, as we'd have to mock `measureText` - not sure if this whole endeavor would make sense - please let me know if you think so

kamilmielnik replied:
Added unit tests for most functions: bd8bb0e7daa8a192e0f114dde071d0c437b52bf4 + 408df899c245c0a396fcadfb8d20f7a9e0405e4d

ranquild commented on frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx:
```diff
@@ -21,17 +21,67 @@ import {
+const getTitleLinesCount = height =>
+  height > TITLE_2_LINES_HEIGHT_THRESHOLD ? 2 : 1;
+
+const formatChangeAutoPrecision = (
```
same here

ranquild commented on frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx:
```diff
@@ -21,17 +21,67 @@ import {
 } from "./SmartScalar.styled";
 
-export default class Smart extends Component {
+const SPACING = parseInt(space(1), 10);
```
I believe we shouldn't rely on `space` returning pixels. I'd rather put an explicit `8` constant here instead

kamilmielnik replied:
aea584d13d4292b7a81bfc86cce3e4567603fee2

ranquild reviewed (commented):
Code-wise it looks like the PR does the right thing, but it adds a lot of computation logic directly to components and makes them a bit hard to follow. I believe we should follow existing conventions, e.g. ones from static-viz, and move all the computation logic to utilities in a separate file. `RowChart` could be used as an example.

We should also request for review from the team who is responsible for viz.

kamilmielnik committed:
Don't assume the "space" function will return values in pixels

kamilmielnik committed:
Move Scalar to its own directory

kamilmielnik committed:
Extract constants.ts and utils.ts from Scalar

kamilmielnik committed:
Move SmartScalar to its own directory

kamilmielnik committed:
Extract constants.ts and utils.ts from SmartScalar

kamilmielnik committed:
Extract styles computation out of SmartScalar

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

kamilmielnik committed:
Extract getValueWidth & getValueHeight to utils

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

kamilmielnik committed:
Update value height computation

kamilmielnik committed:
Update typing

kamilmielnik committed:
Make capitalization work as intended

kamilmielnik committed:
Fix inversed calculation

kamilmielnik committed:
Add missing gridSize prop

kamilmielnik committed:
Remove capitalization logic and improve breakpoints
- I believe the original intention was to capitalize only the first character, but every word is capitalized and it looks weird

kamilmielnik committed:
Extract createCardsRow

kamilmielnik committed:
Break down the tests into more suites because cypress can't consistently handle loading 20 dashboard card simultaneously

kamilmielnik committed:
Improve names

kamilmielnik committed:
Remove mistakenly added size 1 smart cards from tests

WiNloSt commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:
```diff
@@ -145,17 +164,32 @@ export default class Smart extends Component {
+      previousValue,
+      settings.column(column),
     );
+    const disabledSeparator = ""; // avoid creating new translation key
+    const previousValueContent = jt`${disabledSeparator} was ${previousValueDisplay} ${granularityDisplay}`;
```
I'm not sure why we couldn't do the same as `tooltipSeparator` e.g.

```suggestion
    const previousSeparator = (
      <PreviousValueSeparator>•</PreviousValueSeparator>
    );
    const previousValueContent = jt`${previousSeparator} was ${previousValueDisplay} ${granularityDisplay}`;
```

This makes sure the value we use for both tooltip and the value (non-tooltip) uses a similar structure.

kamilmielnik replied:
> why we couldn't do the same as `tooltipSeparator`

Tooltip will always show the separator.
In the card, separator will not be shown if this this text appears in a separate line, which depends on CSS breakpoints. 
Because it depends on CSS breakpoints, we can't use the same styled component in the card and in the tooltip.

kamilmielnik replied:
Alright! Things are much simpler now: 84ae77da84eed1b125f48002e5a788404d673f68

WiNloSt commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:
```diff
@@ -224,23 +262,46 @@ export default class Smart extends Component {
+                    •
+                  </PreviousValueSeparator>
+
+                  {previousValueContent}
```
nit: the tooltip uses
`{jt`${tooltipSeparator} was ${previousValueDisplay} ${granularityDisplay}`}`

but the non-tooltip version extracts those to a variable `previousValueContent`. I think both are similar and should use the same code style. e.g. either inlining both or extracting both to a variable.

kamilmielnik replied:
Answered here: https://github.com/metabase/metabase/pull/31722#discussion_r1245285127
Let's keep the discussion there, I will resolve this comment.

kamilmielnik replied:
64d5d1a71632869e49d0b5a15b99ccc74af25130

WiNloSt commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:
One thing I found confusing about this file is that there are 2 similar items 
1. The previous value in the tooltip
2. The previous value in the card
They kind of reference each other's value but they don't use the same structure when rendering. I think this could be simplified more.

kamilmielnik replied:
Things should be simpler now: 64d5d1a71632869e49d0b5a15b99ccc74af25130 + 84ae77da84eed1b125f48002e5a788404d673f68

kamilmielnik replied:
+ a379593833ad693c920858a023c5df85aa86bb24

WiNloSt reviewed (commented):
This is 80-90% review. I'll review this the first thing tomorrow, as I can't continue now 🤯

ranquild reviewed (approved):
Looks good, however we need to add tests to the extracted utils. Please take a look at comments from @WiNloSt  - I agree that this can be further simplified.

alxnddr commented on e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:
```diff
@@ -0,0 +1,524 @@
+        scalarContainer.then($element => assertIsEllipsified($element[0]));
+        scalarContainer.realHover();
+
+        popover().within(() => {
```
nit: you could write `popover().findByText("18,760");`

kamilmielnik replied:
80af793aa1d5265f5f2d698de112b1922340baa0

kamilmielnik commented:
> Looks good, however we need to add tests to the extracted utils.

@ranquild which utils do you have in mind?

alxnddr reviewed (approved):
I agree with comments from Alex P and Kelvin, besides that looks good

WiNloSt commented on e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:
```diff
@@ -0,0 +1,524 @@
+        ]);
+      });
+
+      it("should truncate value and show value tooltip on hover", () => {
```
Separating tests could make it easier to understand however with the application our size, this increases the test time significantly.
#### This PR which has 3 tests in `1x2 card`
![Screenshot 2023-06-29 at 1 20 52 PM](https://github.com/metabase/metabase/assets/1937582/71bd0acf-dc17-4b62-9e2a-7da01831fb00)

#### If we assert everything in the same test
![Screenshot 2023-06-29 at 1 21 55 PM](https://github.com/metabase/metabase/assets/1937582/614c1eee-0ad6-4aa7-81da-113c748c907c)

Especially when each test doesn't make any changes to the application it's safe to assert different things in the same test to put less burden on our CI runners.

kamilmielnik replied:
> Separating tests could make it easier to understand however with the application our size

"this becomes even more important" is another good ending to that sentence. Yeah, it's a balance.

But it feels wrong to sacrifice quality of tests because of the infrastructure. I'd rather improve the infrastructure.

----

@WiNloSt I didn't get from your comment whether you're suggesting to do this or just making me aware.
Please let me know if you think I should change it.

WiNloSt replied:
I was suggesting you combine all the assertions into a single test, sorry that it wasn't clear. Looping @nemanjaglumac in, since he has the most insights to our CI and he suggested me the same thing.

> But it feels wrong to sacrifice quality of tests because of the infrastructure. I'd rather improve the infrastructure.

you made a good point though.

kamilmielnik replied:
I started doing it but it starts to look very cloudy and I can't follow it anymore:
- Even though I wrote this code, it's hard for me to say where a test begins and ends, especially if there were more than 3 `it` calls
- Variables clash - after changing viewport size we often need to reacquire reference to the DOM element and we'd have to go with `let`
- Tests are supposed to be isolated and independent - the suggestion goes against it

IMO we should not be doing this.

kamilmielnik replied:
> We can revisit the test later once Nemanja has provided some more insights for us.

The bad news is that e2e tests are timing out, so I might do it after all: https://github.com/metabase/metabase/actions/runs/5409602915/jobs/9830378200?pr=31722

WiNloSt replied:
🙃

kamilmielnik replied:
Optimized and added a comment: faadfa0c10132767c887922441fe0cd0b5cb1413
I used `/** */` comments to make test cases more prominent.

WiNloSt replied:
Now we pray to the CI god 🤞

kamilmielnik replied:
In the end I had to adjust the tests because of the optimization:
- after the change, a single test was testing 2 tooltips, so I needed to trigger `mouseleave` to make sure at most 1 tooltip is displayed at a time
- test cases with multiple `cy.viewport()` calls were flaky unless they were wrapped in `it` block, so I wrapped them back

This change made tests very difficult to debug.
I hope we will improve our infrastructure so that we don't need to compromise on quality.

kamilmielnik committed:
Sort props to see differences

kamilmielnik committed:
Fix vertical center alignment in Scalar
- this change also affects SmartScalar, see discussion: https://metaboat.slack.com/archives/C057T1QTB3L/p1688020307409629?thread_ts=1687997714.132239&cid=C057T1QTB3L

kamilmielnik committed:
Replace within() + contains() with findByText()

kamilmielnik committed:
Don't export getTitleHeight
- move it to below where it's used

kamilmielnik committed:
Remove unused gridSize prop

kamilmielnik commented on frontend/src/metabase/visualizations/visualizations/Scalar/utils.ts:
```diff
@@ -0,0 +1,53 @@
+export const getTitleLinesCount = (height: number) =>
+  height > TITLE_2_LINES_HEIGHT_THRESHOLD ? 2 : 1;
+
+export const getTitleHeight = ({
```
33943c19d79ae5aec031e3e02903191e492aa90a

kamilmielnik commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:
```diff
@@ -224,23 +262,46 @@ export default class Smart extends Component {
+                  id="SmartScalar-PreviousValue"
+                  responsive
+                >
+                  <PreviousValueSeparator gridSize={gridSize}>
```
0f9e6015e826ad8712fbabdf288f6e76896ba3b6

kamilmielnik commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.styled.tsx:
```diff
@@ -0,0 +1,73 @@
+`;
+
+export const PreviousValueSeparator = styled(Separator)<{
+  gridSize?: { width: number };
```
0f9e6015e826ad8712fbabdf288f6e76896ba3b6

kamilmielnik committed:
Use consistent code style

kamilmielnik committed:
Simplify separators logic

kamilmielnik committed:
Remove concatRecursively

kamilmielnik commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.ts:
```diff
@@ -0,0 +1,112 @@
+  return availablePreviousValueWidth >= previousValueWidth;
+};
+
+export const concatRecursively = (value: string | string[]) => {
```
a379593833ad693c920858a023c5df85aa86bb24

WiNloSt reviewed (approved):
Looks awesome. We can revisit the test later once Nemanja has provided some more insights for us.

kamilmielnik committed:
Fix typo

kamilmielnik committed:
Optimize test suite

kamilmielnik committed:
Fix typo

kamilmielnik committed:
Add unit tests for Scalar utils: getValueHeight & getValueWidth

kamilmielnik committed:
Add unit tests for SmartScalar utils

kamilmielnik committed:
Fix e2e tests
- update card width in test caes
- wrap cases requiring cy.viewport() in it to prevent test issues

kamilmielnik committed:
Merge branch 'master' into 31628-trend-charts

iethree commented on frontend/src/metabase/visualizations/shared/components/RowChart/RowChart.tsx:
```diff
@@ -48,7 +48,7 @@ export interface RowChartProps<TDatum> {
   tickFormatters?: ChartTicksFormatters;
   labelsFormatter?: (value: NumberValue) => string;
-  measureText: TextMeasurer;
+  measureText: TextWidthMeasurer;
```
we changed the type here, but never the function that was passed, causing https://github.com/metabase/metabase/issues/32044

paoliniluis replied:
why didn't typescript catch this?

iethree replied:
The visualization parent component is still javascript
